klaxer/users.py

```diff
@@ -0,0 +1,172 @@
+    name = Column(String)
+    email = Column(String)
+    registered = Column(Boolean, default=False)
+    approved = Column(Boolean, default=False)
+    signup_date = Column(Date, default=datetime.utcnow())
+    api_key = Column(String, default=None)
+    calls = Column(Integer, default=0)
+    messages = relationship("KlaxerMessage", backref="user")
+
+    def __repr__(self):
+        return '<KlaxerUser {}>'.format(self.id)
+
+    def to_dict(self):
+        """Dump the user object to a `dict`.
+
+        :returns: the user object as a `dict`
+        :rtype: `dict`
+
+        """
+        return {
+            'user_id': self.id,
+            'name': self.name,
+            'email': self.email,
+            'registered': self.registered,
+            'approved': self.approved,
+            'signup_date': self.signup_date,
+            'api_key': self.api_key,
+            'calls': self.calls,
+            'messages': [message.text for message in self.messages]
+        }
+
+    def to_json(self, *args, **kwargs):
+        """Dump the user object to JSON.
+
+        Accepts arguments and keyword arguments for `json.dumps()`.
+        """
+        body = self.to_dict()
+        body['signup_date'] = body['signup_date'].isoformat()
+        return json.dumps(body, *args, **kwargs)
+
+
+class KlaxerMessage(Base):
+    """Message to display to users."""
+    __tablename__ = 'message'
+
+    id = Column(Integer, primary_key=True)
+    text = Column(Text)
+    user_id = Column(Integer, ForeignKey('user.id'))
+    can_dismiss = Column(Boolean, default=True)
+
+    def __repr__(self):
+        return '<KlaxerMessage {}>'.format(self.id)
+
+
+def register(name, email):
+    """Register for an API key with the given name and email.
+
+    :param name: the name of the user/organization
+    :param email: the contact email for the user
+    :returns: an instance the of the user object that was registered
+    :rtype: `klaxer.users.KlaxerUser`
+
+    """
+    user = KlaxerUser(name=name, email=email)
+    user.api_key = uuid4().hex
+    user.registered = True
+    session.add(user)
+    session.commit()
+    return user
+
+
+def create_user(name, email):
+    """Create a user via registration, and add the welcome messages.
+
+    :param name: the name of the user/organization
+    :param email: the contact email for the user
+    :returns: an instance of the user object that was registered
+    :rtype: `klaxer.users.KlaxerUser`
+
+    """
+    user = register(name, email)
+    add_message(user=user, text=config.MSG_WELCOME)
+    add_message(user=user, text=config.MSG_UNVERIFIED, can_dismiss=False)
+    return user
+
+
+def approve(user):
+    """Approve a registered user so that they can use API endpoints.
+
+    :param user: an instance of the `KlaxerUser` object to verify
+    :returns: an instance of the user that matched after updating it
+    :rtype: `klaxer.users.KlaxerUser`
+
+    """
+    user.approved = True
```

Comment:
Do we want to guard against this being called multiple times? Invoked it twice, second time through a `StopIteration` error since the user was already approved.  I can see it going either way.